### ISSUE TITLE
closing a stream followed by respond(0) can error if close is not syncronously putting stream in closed state

### ISSUE THREAD

**Issue by gnoff** [MattiasBuelens/web-streams-polyfill#111]
Hi,

If calling `controller.close` does not always put the stream into closed state (e.g. there are pending pull into requests) is there a way to identify how/when to repond with 0 bytes written to byobRequests that might still be active?

in https://github.com/whatwg/streams/issues/1170 there is a discussion around needing to `.respond(0)` immediately following the close call but since the close call is not necessarily synchronously moving the stream into closed state the `respond(0)` can error.

I found this issue when working with the polyfill v 3.1.1 and was doing some things that aren't recommended like enqueueing chunks before using the view that was provided in a byobRequest. I upgraded to 3.2 which made those enqueues cause the byobRequest to become invalidated which seems to resolve the specific issue I was running into (and I stopped preempting the byob view with enqueued chunks)

However in investigating this issue I can still get myself into a state where a byobRequest exists after calling close and I'd like to know if there is a way to assert we're in closed status before trying to finalize the stream with a last `controller.byobRequest.repond(0)`

**Comment by MattiasBuelens** [comment]
Huh, interesting. I think this might be a spec bug. 🤔 

As you mentioned, you can only call `respond(0)` if the stream is already in the `"closed"` state (step 4 of [ReadableByteStreamControllerRespond](https://streams.spec.whatwg.org/commit-snapshots/9538b77788be6ff7304fcf582823e3d56463ef2a/#readable-byte-stream-controller-respond)). But we only transition to `"closed"` after the *reader* has read the last chunk in the queue. I think we should instead check if `[[closeRequested]]` is true, i.e. if the underlying source is done *putting* new chunks into the queue.

Do you perhaps have a reproduction case that triggers this incorrect behavior? That can help further investigation, and would make for a good test to go along with the spec change. 🙂

**Comment by MattiasBuelens** [comment]
Thanks for the repro!

> I explain in the comments that it is entirely possible this edge case is impossible with the changes implemented in 3.2 however I don't know the spec well enough to know if there are other edge cases that could lead to a similar state occurring that lead to this error.

Yes, it does look like this was fixed in whatwg/streams#1171, which I implemented in #103 and shipped in version 3.2.0.

[I thought](https://github.com/MattiasBuelens/web-streams-polyfill/issues/111#issuecomment-1067416386) there might still be a situation where you call `respond(0)` when `closeRequested` is true *without* `state` also being `"closed"`, but I can't seem to make that happen. The stream will always first try to fill a read request from the queue, and will only call `pull()` (and create a BYOB request) when it has completely drained the queue. So there cannot be an active BYOB request *while the queue is non-empty*. Therefore, when you do `controller.close()` while there is an active BYOB request, `queueTotalSize` *must* be 0 and [ReadableByteStreamControllerClose](https://streams.spec.whatwg.org/commit-snapshots/9538b77788be6ff7304fcf582823e3d56463ef2a/#readable-byte-stream-controller-close) *must* reach step 6 and transition to the `"closed"` state. Then, when you do `byobRequest.respond(0)`, the stream will have the correct state already and we don't really have to check `closeRequested` in [ReadableByteStreamControllerRespond](https://streams.spec.whatwg.org/commit-snapshots/9538b77788be6ff7304fcf582823e3d56463ef2a/#readable-byte-stream-controller-respond).

Long story short: I don't think we need a spec change after all, since it's already fixed. 😁

**Comment by gnoff** [comment]
👏 I suspected that might be the case. glad to see you could confirm it. my hunch was not backed up by an understanding of the dependent states that might conspire to prevent that state from being possible 😄
